File: CivitasJava/civitas/Jugador.java
package civitas;

import java.util.ArrayList;
import GUI.Dado;

public class Jugador implements Comparable<Jugador> {

    // ATRIBUTOS

    protected static int CasasMax = 4;
    protected static int CasasPorHotel = 4;
    protected boolean encarcelado;
    protected static int HotelesMax = 4;
    private String nombre;
    private int numCasillaActual;
    protected static float PasoPorSalida = 1000;
    protected static float PrecioLibertad = 200;
    private boolean puedeComprar;
    private float saldo;
    private static float SaldoInicial = 7500;
    private Sorpresa_SalirCarcel salvoconducto;
    private ArrayList<TituloPropiedad> propiedades;
    protected Boolean especulador;

    // CONSTRUCTORES

    Jugador(String name) {
        this.nombre = name;
        this.saldo = Jugador.SaldoInicial;
        this.encarcelado = false;
        this.puedeComprar = true;
        this.salvoconducto = null;
        this.propiedades = new ArrayList<>();
        this.especulador=false;
    }

    protected Jugador(Jugador otro) {
        this.nombre = otro.getNombre();
        this.saldo = otro.getSaldo();
        this.encarcelado = otro.isEncarcelado();
        this.puedeComprar = otro.getPuedeComprar();
        this.propiedades = otro.getPropiedades();
        this.salvoconducto = otro.salvoconducto;
        this.especulador=otro.especulador;
    }

    // METODOS GET Y SET

    protected int getCasasMax() {
        return CasasMax;
    }
    
    public boolean isEspeculador()
    {
        return this.especulador;
    }

    int getCasasPorHotel() {
        return CasasPorHotel;
    }

    protected int getHotelesMax() {
        return HotelesMax;
    }

    public String getNombre() {
        return nombre;
    }

    int getNumCasillaActual() {
        return numCasillaActual ;
    }

    private float getPrecioLibertad() {
        return PrecioLibertad;
    }

    private float getPremioPasoSalida() {
        return PasoPorSalida;
    }

    public ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }

    public ArrayList<String> getPropiedadesString() {
        ArrayList<String> prop = new ArrayList<>();

        for (int i = 0; i < this.propiedades.size(); i++) {
            prop.add(propiedades.get(i).getNombre());
        }

        return prop;
    }

    boolean getPuedeComprar() {
        return puedeComprar;
    }

    public float getSaldo() {
        return saldo;
    }

    // METODOS

    boolean cancelarHipoteca(int ip) {
        boolean result = false;
        if (!this.encarcelado) {
            if (this.existeLaPropiedad(ip)) {
                TituloPropiedad propiedad = this.propiedades.get(ip);
                float cantidad = propiedad.getImporteCancelarHipoteca();
                boolean puedoGastar = this.puedoGastar(cantidad);
                if (puedoGastar) {
                    result = propiedad.cancelarHipoteca(this);
                    if (result) {
                        Diario.getInstance().ocurreEvento(
                                "El jugador " + this.nombre + " cancela la hipoteca de la propiedad " + ip + "\n");
                    }
                }
            }
        }
        return result;
    }

    int cantidadCasasHoteles() {

        int sal = 0;

        for (int i = 0; i < this.propiedades.size(); i++) {
            sal += propiedades.get(i).cantidadCasasHoteles();
        }

        return sal;
    }

    @Override
    public int compareTo(Jugador otro) {

        return Float.compare(this.getSaldo(), otro.getSaldo());
    }

    boolean comprar(TituloPropiedad titulo) {
        boolean result = false;
        if (!this.encarcelado) {
            if (this.puedeComprar) {
                float precio = titulo.getPrecioCompra();
                if (this.puedoGastar(precio)) {
                    result = titulo.comprar(this);
                    if (result) {
                        this.propiedades.add(titulo);
                        Diario.getInstance().ocurreEvento(
                                "El jugador " + this.nombre + " compra la propiedad " + titulo.toString()+ "\n");
                    }
                    this.puedeComprar = false;
                }
            }
        }
        return result;
    }

    boolean construirCasa(int ip) {
        boolean result = false;
        boolean puedoEdificarCasa = false;
        if (!this.encarcelado) {
            boolean existe = this.existeLaPropiedad(ip);
            if (existe) {
                TituloPropiedad propiedad = this.propiedades.get(ip);
                puedoEdificarCasa = this.puedoEdificarCasa(propiedad);
                if (puedoEdificarCasa) {
                    result = propiedad.construirCasa(this);
                    if (result) {
                        Diario.getInstance()
                                .ocurreEvento("El jugador " + nombre + " construye casa en la propiedad " + ip+ "\n");
                    }
                }

            }
        }
        return result;
    }

    boolean construirHotel(int ip) {
        boolean result = false;
        if (!this.encarcelado) {
            if (this.existeLaPropiedad(ip)) {
                TituloPropiedad propiedad = this.propiedades.get(ip);
                boolean puedoEdificarHotel = this.puedoEdificarHotel(propiedad);
                if (puedoEdificarHotel) {
                    result = propiedad.construirHotel(this);
                    int casasPorHotel = this.getCasasPorHotel();
                    propiedad.derruirCasas(casasPorHotel, this);
                    Diario.getInstance()
                            .ocurreEvento("El jugador " + nombre + " construye hotel en la propiedad " + ip+ "\n");

                }
            }
        }
        return result;
    }

    protected boolean debeSerEncarcelado() {

        boolean carcel = false;

        if (!this.encarcelado) {
            if (!this.tieneSalvoConducto())
                carcel = true;
            else
                this.perderSalvoConducto();
        }

        return carcel;
    }

    boolean enBancarrota() {

        return this.getSaldo() <= 0;
    }

    boolean encarcelar(int numCasillaCarcel) {

        if (this.debeSerEncarcelado()) {
            this.moverACasilla(numCasillaCarcel);
            this.encarcelado = true;
            Diario.getInstance().ocurreEvento("El jugador " + this.getNombre() + " ha sido encarcelado.\n");
        }

        return this.encarcelado;
    }

    private boolean existeLaPropiedad(int ip) {

        return ip >= 0 && ip < propiedades.size();
    }

    boolean hipotecar(int ip) {
        boolean result = false;
        if (!this.encarcelado) {
            if (this.existeLaPropiedad(ip)) {
                TituloPropiedad propiedad = this.propiedades.get(ip);
                result = propiedad.hipotecar(this);
            }
            if (result) {
                Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " hipoteca la propiedad " + ip+ "\n");
            }
        }
        return result;
    }

    public boolean isEncarcelado() {

        return this.encarcelado;
    }

    boolean modificarSaldo(float cantidad) {

        this.saldo += cantidad;

        Diario.getInstance().ocurreEvento(
                "Se ha modificado el saldo del jugador " + this.getNombre() + " en: " + cantidad + " unidades.\n");

        return true;
    }

     boolean moverACasilla(int numCasilla) {

        boolean ret = false;

        if (!this.isEncarcelado()) {
            this.numCasillaActual = numCasilla;
            this.puedeComprar = false;
            Diario.getInstance().ocurreEvento("El jugador " + this.getNombre() + " se mueve hasta la casilla " + numCasilla+ "\n");
            ret = true;
        }

        return ret;
    }

    boolean obtenerSalvoConducto(Sorpresa_SalirCarcel sorpresa) {

        if (this.encarcelado)
            this.salvoconducto = sorpresa;

        return (!this.encarcelado);
    }

    boolean paga(float cantidad) {

        return modificarSaldo(cantidad * -1);

    }

    boolean pagaAlquiler(float cantidad) {

        return this.pagaImpuesto(cantidad);
    }

    boolean pagaImpuesto(float cantidad) {

        return this.encarcelado ? false : this.paga(cantidad);

    }

    boolean pasaPorSalida() {

        this.modificarSaldo(Jugador.PasoPorSalida);
        Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " pasa por la salida y se le incrementa el saldo en: " + Jugador.PasoPorSalida + " rupias.\n");
        return true;
    }

    protected void perderSalvoConducto() {

        this.salvoconducto.usada();
        this.salvoconducto = null;
    }

    boolean puedeComprarCasilla() {
        
        this.puedeComprar = !this.encarcelado;
        return this.puedeComprar;
    }

    private boolean puedeSalirCarcelPagando() {

        return this.getSaldo() >= Jugador.PrecioLibertad;
    }

    private boolean puedoEdificarCasa(TituloPropiedad propiedad) {
        return this.getSaldo() >= propiedad.getPrecioEdificar();
    }

    private boolean puedoEdificarHotel(TituloPropiedad propiedad) {
        boolean puedoEdificarHotel = false;
        float precio = propiedad.getPrecioEdificar();
        if (this.puedoGastar(precio)) {
            if (propiedad.getNumHoteles() < this.getHotelesMax()) {
                if (propiedad.getNumCasas() >= this.getCasasPorHotel()) {
                    puedoEdificarHotel = true;
                }
            }
        }
        return puedoEdificarHotel;
    }

    private boolean puedoGastar(float precio) {

        return this.encarcelado ? false : this.saldo >= precio;

    }

    boolean recibe(float cantidad) {

        boolean ret = false;
        if (!this.isEncarcelado()) {
            ret = this.modificarSaldo(cantidad);
        }

        return ret;
    }

    boolean salirCarcelPagando() {

        boolean salida = false;

        if (this.encarcelado && this.puedeSalirCarcelPagando()) {

            this.paga(Jugador.PrecioLibertad);
            this.encarcelado = false;
            salida = true;
            Diario.getInstance().ocurreEvento("El jugador " + this.nombre + " paga para salir de la carcel.\n");

        }

        return salida;

    }

    boolean salirCarcelTirando() {

        if (Dado.getInstance().salgoDeLaCarcel()) {
            this.encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador " + this.getNombre() + " sale de la carcel.\n");
        }

        return this.isEncarcelado();
    }

    boolean tieneAlgoQueGestionar() {

        return this.propiedades.size() > 0;

    }

    boolean tieneSalvoConducto() {

        return this.salvoconducto != null;

    }

    @Override
    public String toString() {

        String salida;
        salida = "Nombre: " + this.getNombre() + "\n" + "Saldo: " + this.getSaldo() + "\n" + "Casilla actual: "
                + this.getNumCasillaActual() + "\n" + "Encarcelado: " + this.isEncarcelado()+"\n";

        return salida;
    }

    boolean vender(int ip) {

        boolean ret = false;

        if (!this.isEncarcelado() && this.existeLaPropiedad(ip)) {
            if (this.propiedades.get(ip).vender(this)) {
                this.propiedades.remove(ip);
                Diario.getInstance().ocurreEvento("Se ha vendido una propiedad.\n");
                ret = true;
            }
        }

        return ret;
    }
}

File: CivitasRuby/Jugador.rb
module Civitas
    include Comparable
    class Jugador
        @@CasasMax=4
        @@CasasPorHotel=4
        @@HotelesMax=4
        @@PasoPorSalida=1000
        @@PrecioLibertad=200    
        @@SaldoInicial=7500
        
        attr_reader :CasasMax, :CasasPorHotel, :HotelesMax, :nombre, :numCasillaActual, :PrecioLibertad, :encarcelado
        attr_reader :PremioPasoSalida, :propiedades, :puedeComprar, :saldo, :PasoPorSalida
        attr_accessor :saldo, :salvoconducto, :puedeComprar, :encarcelado, :propiedades, :numCasillaActual
        
        def initialize(n)
            @nombre=n
            @numCasillaActual=0
            @encarcelado=false
            @puedeComprar=true
            @saldo=@@SaldoInicial
            @propiedades=[]
            @salvoconducto=nil
        end

        def copia(otro)

            @nombre=otro.nombre

            @saldo = otro.saldo
            @numCasillaActual = otro.numCasillaActual
            @salvoconducto = otro.salvoconducto
            @puedeComprar = otro.puedeComprar
            @encarcelado = otro.encarcelado
            @propiedades = otro.propiedades

        end

        def debeSerEncarcelado 
            @encarcelado ? ret=false : ret= true 

            if !@encarcelado&&!tieneSalvoConducto
                ret=true
            else 
                if tieneSalvoConducto
                    perderSalvoConducto
                    Diario.instance.ocurre_evento('El jugador ' + @nombre + ' se libra de la carcel.')
                    ret=false
                end
            end

            return ret
        end

        def encarcelar(numCasillaCarcel)
            
            if debeSerEncarcelado
                moverACasilla(numCasillaCarcel)
                @encarcelado=true
                Diario.instance.ocurre_evento('El jugador ' + @nombre + ' ha sido encarcelado.')
            end

            return @encarcelado
        end

        def obtenerSalvoconducto(s)
            
            ret=false

            if !encarcelado
                ret=true
                @salvoconducto=s
            end
            
            return ret
        end

        def perderSalvoConducto

            @salvoconducto.usada()

        end

        def tieneSalvoConducto

            return @salvoconducto!=nil

        end

        def puedeComprarCasilla

            @encarcelado ? @puedeComprar=false : @puedeComprar=true
            return @puedeComprar

        end

        def paga(cantidad)
        
            return modificarSaldo(-1*cantidad)
            
        end

        def pagaImpuesto(cantidad)
        
            return @encarcelado ? true : paga(cantidad)
        
        end

        def pagaAlquiler(cantidad)

            return pagaImpuesto(cantidad)

        end

        def recibe(cantidad)
        
            return encarcelado ? false : modificarSaldo(cantidad)
        
        end
        
        def modificarSaldo(cantidad)
        
            @saldo+=cantidad

            Diario.instance.ocurre_evento('Se ha modificado el saldo del jugador ' + @nombre + ' en: ' + cantidad.to_s + ' unidades.')
            
            return true
        
        end

        def moverACasilla(numCasilla)
        
            ret=false

            if !@encarcelado
                @numCasillaActual=numCasilla
                @puedeComprar=false
                ret=true
                Diario.instance.ocurre_evento('El jugador ' + @nombre + ' se mueve hasta la casilla ' + numCasilla.to_s)
            end

            return ret
        end


        def cancelarHipoteca(ip)

            ret=false

            if !@encarcelado
                
                if existeLaPropiedad(ip)

                    propiedad=@propiedades[ip]
                    importe=propiedad.getImporteCancelarHipoteca
                    
                    if puedoGastar(importe)
                        res=propiedad.cancelarHipoteca(self)

                        if res  
                            Diario.instance.ocurre_evento("El jugador "+ @nombre + " cancela la hipoteca de la propiedad " + ip.to_s)
                        end
                    end
                end
            end

            return ret 
        end

        def cantidadCasasHoteles
            
            veces=@propiedades.length
            i=0
            numCasasHoteles=0
            veces.times do
                numCasasHoteles+=@propiedades[i].numCasas
                numCasasHoteles+=@propiedades[i].numHoteles
                i+=1
            end
            
            return numCasasHoteles

        end

        def compareTo(otro)

            return self<=>otro  #NOS DEVUELVE -1 SI SELF ES MAYOR QUE EL OTRO
                                #NOS DEVUELVE 0 SI SON IGUALES
                                #NOS DEVUELVE 1 SI OTRO ES MAYOR QUE SELF

        end
        
        def <=> (other)

            @saldo<=>other.saldo

        end
        
        def comprar(titulo)

            ret=false

            if !@encarcelado

                if @puedeComprar

                    precio=titulo.precioCompra

                    if puedoGastar(precio)
                    
                        ret=titulo.comprar(self)

                        if ret
                            @propiedades<<titulo
                            Diario.instance.ocurre_evento("El jugador "+ @nombre + " compra la propiedad " + titulo.toString)
                        end
                        
                        @puedeComprar=false
                    end
                end
            end

            return ret
        end
        
        def construirCasaJug(ip) 

            res=false
            edificar=false

            if !@encarcelado 
                existe=existeLaPropiedad(ip)
                if existe
                    propiedad=@propiedades[ip]
                    edificar=puedoEdificarCasa(propiedad)
                    if edificar
                        res=propiedad.construirCasaProp(self)
                        
                        if res
                            Diario.instance.ocurre_evento("El jugador "+ @nombre + "construye casa en la propiedad"+ip.to_s)
                        end
                    end
                end
            end

            return res
        end
        
        def construirHotelJug(ip)

            ret=false
            edificar=false

            if !@encarcelado
                
                if existeLaPropiedad(ip)

                    propiedad=@propiedades[ip]
                    edificar=puedoEdificarHotel(propiedad)

                    if edificar

                        ret=propiedad.construirHotelProp(self)
                        casasPorHotel=@@CasasPorHotel
                        propiedad.derruirCasas(casasPorHotel, self)
                        Diario.instance.ocurre_evento("El jugador "+@nombre+" construye hotel en la propiedad "+ip.to_s)
                    end
                end
            end

            return ret
        end

        def getPropiedadesString
            prop=[]

            for i in (0..propiedades.size-1)
                prop<<propiedades[i].nombre
            end

            return prop
        end 
        
        def enBancarrota

            return @saldo <= 0

        end
    
        def existeLaPropiedad(ip)
            
            return ip >= 0 && ip < propiedades.length

        end

        def hipotecar(ip)

            res=false

            if !@encarcelado

                if existeLaPropiedad(ip)
                    propiedad=@propiedades[ip]
                    res=propiedad.hipotecar(self)
                end

                if res
                    Diario.instance.ocurre_evento("El jugador "+@nombre+" hipoteca la propiedad "+ip.to_s)
                end
            end
        end

        def isEncarcelado
        
            return @encarcelado

        end

        def pasaPorSalida
            
            modificarSaldo(@@PasoPorSalida)
            Diario.instance.ocurre_evento('El jugador ' + @nombre + ' pasa por la salida y se le incrementa el saldo en: ' + @@PasoPorSalida.to_s + ' unidades.')

        end
        
        def puedeSalirCarcelPagando

            return @saldo>=@@PrecioLibertad

        end

        def puedoEdificarCasa(propiedad)
            
            precio=propiedad.precioEdificar
            puedoEdificarCasa=false

            if puedoGastar(precio) && propiedad.numCasas < @@CasasMax
                puedoEdificarCasa=true;
            end

            return puedoEdificarCasa
        end

        def puedoEdificarHotel(propiedad) 

            puedoEdificarHotel=false
            precio=propiedad.precioEdificar

            if puedoGastar(precio)
            
                if (propiedad.numCasas >= @@CasasPorHotel) && (propiedad.numHoteles <  @@HotelesMax)
                    puedoEdificarHotel=true
                end
            end

            return puedoEdificarHotel
        end

        def puedoGastar(precio)
            
            ret=false

            if !@encarcelado 
                ret= @saldo>=precio
            end

            return ret
        end

        def salirCarcelPagando

            ret=false

            if @encarcelado && puedeSalirCarcelPagando
                paga(@@PrecioLibertad)
                @encarcelado=false
                Diario.instance.ocurre_evento('El jugador ' + @nombre + ' paga para salir de la carcel.')
                ret=true
            end

            return ret
        end

        def salirCarcelTirando
            
            ret=Dado.instance.salgoDeLaCarcel  

            if ret
                @encarcelado=false
                Diario.instance.ocurre_evento('El jugador ' + @nombre + ' consigue salir tirando de la carcel.')
            end

            return ret

        end
        
        def tieneAlgoQueGestionar

            return @propiedades.length!=0

        end

        def toString

            salida= 'Nombre: ' + @nombre + "\n"
            salida+= 'Saldo: ' + @saldo.to_s + "\n"
            salida+= 'Casilla actual: ' + @numCasillaActual.to_s + "\n"
            salida+= 'Encarcelado: ' + isEncarcelado.to_s + "\n\n"

            return salida
        
        end
        
        def vender(ip)

            ret=false

            if !@encarcelado
                if existeLaPropiedad(ip)
                    if @propiedades[ip].vender(self)
                        @propiedades.delete_at(ip)
                        ret=true
                        Diario.instance.ocurre_evento('Se ha vendido una propiedad.')
                    end
                end
            end

            return ret

        end
         
        #PROTECTED Y PRIVADO DE LOS CONSULTORES

        private :CasasMax, :HotelesMax, :PrecioLibertad, :PremioPasoSalida
        
        #PROTECTED Y PRIVADO DE LOS METODOS

        private :existeLaPropiedad, :perderSalvoConducto, :puedeSalirCarcelPagando, :puedoEdificarCasa
        private :puedoEdificarHotel, :puedoGastar
        protected :debeSerEncarcelado
    end
end
